core/frontend/src/BackgroundMapGeometry.ts

```diff
@@ -124,6 +124,15 @@ export class BackgroundMapGeometry {
     return Promise.all(promises);
   }
 
+  public async dbToWGS84CartographicFromGcs(db: XYAndZ[]): Promise<Cartographic[]> {
+    const scratch = new Point3d();
+    const promises = db.map(async (p) => {
+      return this.cartesianRange.containsPoint(Point3d.createFrom(p, scratch)) ? this._iModel.spatialToWGS84Cartographic(p) : this.dbToCartographic(p);
```

Comment:
Your usage of `scratch` is problematic - you're wrapping up a bunch of promises that all refer to the same `Point3d` object. It might "just happen" to work because spatialToWGS84Cartographic and dbToCartographic clone it (I didn't check), but that'd be a nasty bug to have to untangle.

Reply:
Probably better if spatialToWGS84ToCartographic takes an array of XYAndZ, and you split the input array into two, producing at most 2 promises.

Reply:
1. So how `dbToWGS84CartographicFromGcs` is worst than  `dbToCartographicFromGcs` ?
2. I dont understand your suggestion... what about moving `scratch` inside the function passed to` .map` ?

Reply:
1. It's not - it suffers the same potential bug.
2. Resolves potential Point3d bug and makes 2 lists:
```
const reproject: XYAndZ[] = [];
const noReproject: XYAndZ[] = [];
const scratch = new Point3d();
for (const p of db) {
  Point3d.createFrom(p, scratch);
  const list = this.cartesianRange.containsPoint(scratch) ? reproject : noReproject;
  list.push(p);
}

// Make promises out of the 2 lists (ignoring empty lists)
```

Reply:
1. nvm, I misread the usage of `scratch` - it's not passed to async functions, `p` is.